constraints/src/main/java/gov/nasa/jpl/aerie/constraints/model/LinearProfile.java

```diff
@@ -1,203 +1,134 @@
-   */
-  private static <T> List<T> processIntersections(final LinearProfile left, final LinearProfile right, final BiFunction<LinearProfilePiece, LinearProfilePiece, T> processor, Interval bounds) {
-    if (left.profilePieces.isEmpty() || right.profilePieces.isEmpty()) return new ArrayList<>();
-
-    // Setup to step through profiles simultaneously,
-    // finding all windows satisfying condition within the supplied bounds
-    final var processedIntersections = new ArrayList<T>();
-    final var rightIter = right.profilePieces.iterator();
-
-    // Loop through the left pieces, stepping up the right pieces
-    // appropriately each time to get intersections
-    if (!rightIter.hasNext()) return processedIntersections;
-    var rightPiece = rightIter.next();
-    for (final var leftPiece : left.profilePieces) {
-      //if left piece ends before bounds start, skip left piece
-      if (profileOutsideBounds(leftPiece, bounds)) continue;
-      // If left piece ends before right piece, skip left piece
-      if (Interval.compareEndToStart(leftPiece.interval, rightPiece.interval) < 0) continue;
-
-      // Step through right pieces ending before left piece starts
-      // If no right pieces ending after left piece starts exist, end the loop
-      while (Interval.compareEndToStart(rightPiece.interval, leftPiece.interval) < 0) {
-        if (rightIter.hasNext()) {
-          rightPiece = rightIter.next();
+    public Windows changePoints() {
+      final var result = IntervalMap.<Boolean>builder().set(this.profilePieces.map(LinearEquation::changing));
+      for (int i = 0; i < this.profilePieces.size(); i++) {
+        final var segment = this.profilePieces.get(i);
+        final var startTime = segment.interval().start;
+        if (i == 0) {
+          if (!segment.interval().contains(Duration.MIN_VALUE)) {
+            result.unset(Interval.at(segment.interval().start));
+          }
         } else {
-          break;
-        }
-      }
-
-      // Process all intersections with right pieces that start before the left piece ends
-      // If we run out of right pieces, end the loop
-      while (Interval.compareStartToEnd(rightPiece.interval, leftPiece.interval) <= 0) {
-        processedIntersections.add(processor.apply(leftPiece, rightPiece));
-
-        // Only step passed right piece if it doesn't exceed the left piece
-        if (Interval.compareEndToEnd(rightPiece.interval, leftPiece.interval) <= 0) {
-          if (rightIter.hasNext()) {
-            rightPiece = rightIter.next();
-            continue;
+          final var previousSegment = this.profilePieces.get(i-1);
+
+          if (Interval.meets(previousSegment.interval(), segment.interval())) {
+            if (previousSegment.value().valueAt(startTime) != segment.value().valueAt(startTime)) {
+              result.set(Interval.at(segment.interval().start), true);
+            }
+          } else {
+            result.unset(Interval.at(segment.interval().start));
```

Comment:
Still trying to wrap my head around what `changePoints` means in the presence of gaps, but it does seem strange to me that we're unsetting `segment.interval().start` but not `previousSegment.interval().end`. Thoughts?

Reply:
gaps mean "unknown" value. So if a segment is surrounded by gaps, we don't know if the first point of the segment is different from the point before, but we do know if the last point of the segment is different from the point before.

Reply:
Ah, I see - the key definition I was missing was the definition of changePoint:
> a changePoint is defined as any point whose value is known to be different from the point before.

Through that lens, this method makes sense :+1: (I was thinking maybe it was a point where a change occurs either before or after, and was getting tied up in a knot. I was also wondering whether the 'clusivity mattered....). 

Perhaps a comment somewhere with this definition would help? Either in the constraints documentation, or just a javadoc comment either on this method, or on the `Profile` interface.